doc/http-server.1
.TH http-server 1 "April 2020" GNU "http-server man page"

.SH NAME
http-server \- a simple zero-configuration command-line http server

.SH SYNOPSIS  
.B http-server
[\fIPATH\fR]
[\fIOPTIONS\fR]

.SH DESCRIPTION
\fBhttp-server\fR is a simple, zero-configuration command-line http server. It is powerful enough for production usage, but it's simple and hackable enough to be used for testing, local development, and learning.

.SH OPTIONS

.TP
.BI [\fIPATH\fR]
The directory to serve.
Defaults to ./public if it exists, and ./ otherwise.

.TP
.BI \-p ", " \-\-port " " \fIPORT\fR
Port to use.
Default is 8080.

.TP
.BI \-a " " \fIADDRESS\fR
Address to use.
Default is 0.0.0.0.

.TP
.BI \-d
Show directory listings.
Default is true.

.TP
.BI \-i
Display autoIndex.
Default is true.

.TP
.BI \-g ", " \-\-gzip
Serve gzip files when possible.
Default is false.

.TP
.BI \-b ", " \-\-brotli
Serve brotli files when possible.
If both brotli and gzip are enabled, brotli takes precedence.
Default is false.

.TP
.BI \-e ", " \-\-ext " " \fIEXTENSION\fR
Default file extension is none is provided.

.TP
.BI \-s ", " \-\-silent
Suppress log messages from output.

.TP
.BI \-\-cors " " [\fIHEADERS\fR]
Enable CORS via the "Access-Control-Allow-Origin" header.
Optionally provide CORS headers list separated by commas.

.TP
.BI \-o " " [\fIPATH\fR]
Open default browser window after starting the server.
Optionally provide a URL path to open the browser window to.

.TP
.BI \-c " " \fITIME\fR
Cache time (max-age) in seconds.
To disable caching, use \-c \-1.
Default is 3600.

.TP
.BI \-U ", " \-\-utc
Use UTC time format in log messages.

.TP
.BI \-\-log\-ip
Enable logging of the client IP address.

.TP
.BI \-P ", " \-\-proxy
Fallback proxy if the request cannot be resolved.

.TP
.BI \-\-username " " \fIUSERNAME\fR
Username for basic authentication.
Can also be specified with the environment variable NODE_HTTP_SERVER_USERNAME.
Defaults to none.

.TP
.BI \-\-password " " \fIPASSWORD\fR
Password for basic authentication.
Can also be specified with the environment variable NODE_HTTP_SERVER_PASSWORD.
Defaults to none.

.TP
.BI \-S ", " \-\-ssl
Enable https.

.TP
.BI \-C ", " \-\-cert " " [\fIFILE\fR]
Path to SSL certificate file.
If not specified, uses cert.pem.

.TP
.BI \-K ", " \-\-key " " [\fIFILE\fR]
Path to SSL key file.
If not specified, uses key.pem.

.TP
.BI \-r ", " \-\-robots " " [\fIUSER\-AGENT\fR]
Respond to /robots.txt request.
If not specified, uses "User-agent: *\\nDisallow: /]"

.TP
.BI \-\-no\-dotfiles
Do not show dotfiles.

.TP
.BI \-h ", " \-\-help
Print usage and exit.

.TP
.BI \-v ", " \-\-version
Print version and exit.

.SH FILES
.B index.html
will be served as the default file to any directory requests.

.B 404.html
will be served if a file is not found. This can be used for SPA hosting to serve the entry page.

.SH COPYING
Copyright (c) 2011-2020 Charlie Robbins, Marak Squires, and the Contributors.
Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

.SH VERSION
Version 0.12.2


lib/http-server.js
'use strict';

var fs = require('fs'),
    union = require('union'),//流中间件
    ecstatic = require('ecstatic'),//静态文件服务器中间件
    httpProxy = require('http-proxy'),//支持websockets的HTTP可编程代理库。它适合于实现反向代理和负载平衡器等组件。
    auth = require('basic-auth'),// 通用的基本授权头字段解析器
    corser = require('corser'),// CORS中间件
    path = require('path'),
    secureCompare = require('secure-compare');// 安全比较算法

// a hacky and direct workaround to fix https://github.com/http-party/http-server/issues/525
function getCaller() {
  try {
    var stack = new Error().stack;
    var stackLines = stack.split('\n');
    var callerStack = stackLines[3];
    return callerStack.match(/at (.+) \(/)[1];
  }
  catch (error) {
    return '';
  }
}

var _pathNormalize = path.normalize;//规范化给定的 path，解析 '..' 和 '.' 片段
path.normalize = function (p) {
  var caller = getCaller();
  var result = _pathNormalize(p);
  // https://github.com/jfhbrook/node-ecstatic/blob/master/lib/ecstatic.js#L20
  if (caller === 'decodePathname') {
    result = result.replace(/\\/g, '/');
  }
  return result;
};

//
// Remark: backwards compatibility for previous
// case convention of HTTP
//
exports.HttpServer = exports.HTTPServer = HttpServer;

/**
 * Returns a new instance of HttpServer with the
 * specified `options`.
 */
exports.createServer = function (options) {
  return new HttpServer(options);
};

/**
 * Constructor function for the HttpServer object
 * which is responsible for serving static files along
 * with other HTTP-related features.
 */
function HttpServer(options) {
  options = options || {};

  if (options.root) {
    this.root = options.root;
  }
  else {
    try {
      fs.lstatSync('./public');// 异步地创建目录
      this.root = './public';
    }
    catch (err) {
      this.root = './';
    }
  }

  this.headers = options.headers || {};

  this.cache = (
    options.cache === undefined ? 3600 :
    // -1 is a special case to turn off caching.
    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Preventing_caching
    options.cache === -1 ? 'no-cache, no-store, must-revalidate' :
    options.cache // in seconds.
  );
  this.showDir = options.showDir !== 'false';
  this.autoIndex = options.autoIndex !== 'false';
  this.showDotfiles = options.showDotfiles;
  this.gzip = options.gzip === true;
  this.brotli = options.brotli === true;
  console.log(options.ext);
  if (options.ext) {//如果有默认文件扩展名
    this.ext = options.ext === true
      ? 'html'
      : options.ext;
  }
  this.contentType = options.contentType ||
    this.ext === 'html' ? 'text/html' : 'application/octet-stream';//浏览器检bai测文件类型，du有两种响应：第一种是MIME（多zhi功能Internet 邮件扩dao充服务，最早zhuan用于邮件系统，后shudao拓展到浏览器中）；另一种，当浏览器无法确定文件类型时，就是application/octet-stream类型。

  var before = options.before ? options.before.slice() : [];

  if (options.logFn) {//如果有日志输出函数
    before.push(function (req, res) {
      options.logFn(req, res);
      res.emit('next');
    });
  }

  if (options.username || options.password) {//如果有用户名和密码
    before.push(function (req, res) {
      var credentials = auth(req);
      console.log(`credentials:${credentials}`);
      // We perform these outside the if to avoid short-circuiting and giving
      // an attacker knowledge of whether the username is correct via a timing
      // attack.
      if (credentials) {
        // if credentials is defined, name and pass are guaranteed to be string
        // type
        var usernameEqual = secureCompare(options.username.toString(), credentials.name);
        var passwordEqual = secureCompare(options.password.toString(), credentials.pass);
        if (usernameEqual && passwordEqual) {
          return res.emit('next');
        }
      }

      res.statusCode = 401;
      res.setHeader('WWW-Authenticate', 'Basic realm=""');
      res.end('Access denied');
    });
  }

  if (options.cors) {//如果允许跨域
    this.headers['Access-Control-Allow-Origin'] = '*';
    this.headers['Access-Control-Allow-Headers'] = 'Origin, X-Requested-With, Content-Type, Accept, Range';
    if (options.corsHeaders) {
      options.corsHeaders.split(/\s*,\s*/)
          .forEach(function (h) { this.headers['Access-Control-Allow-Headers'] += ', ' + h; }, this);
    }
    before.push(corser.create(options.corsHeaders ? {
      requestHeaders: this.headers['Access-Control-Allow-Headers'].split(/\s*,\s*/)
    } : null));
  }

  if (options.robots) {//如果允许响应 robots.txt
    before.push(function (req, res) {
      if (req.url === '/robots.txt') {
        res.setHeader('Content-Type', 'text/plain');
        var robots = options.robots === true
          ? 'User-agent: *\nDisallow: /'
          : options.robots.replace(/\\n/, '\n');

        return res.end(robots);
      }

      res.emit('next');
    });
  }

  /**
   * 自定义中间件
   */
  const root = this.root;
  before.push(async function(req,res){ 
    const host = req.headers.host;
    const {domains=[],appDir,customization={},defaultDir={}} = global.config;
    const appName=req.url&&req.url.match(/^\/(.*?)\//)?req.url.match(/^\/(.*?)\//)[1]:'';
    const appFolder = appDir[appName]||appDir['mis'];//子应用文件夹路径
    const matchDomain = domains.find(({domain})=>host === domain);
    const defaultPath =  defaultDir.main + appFolder + req.url;// 默认路径
    if(matchDomain){//如果当前域名（仲裁委）有定制
      const acFolder = '/' + matchDomain.name;//仲裁委文件夹路径
      const customPath = defaultDir.custom + acFolder + appFolder + req.url;// 定制的路径
      await new Promise((resolve) => {
          fs.access(path.join(root, customPath), fs.constants.F_OK, (err) => {
              req.url = err?defaultPath:customPath;
              resolve();
          });
      });
    }else{//否则没有定制
      req.url = defaultPath;
    }
    console.log(req.url);
    res.emit('next');
  });


  before.push(ecstatic({
    root: this.root,
    cache: this.cache,
    showDir: this.showDir,
    showDotfiles: this.showDotfiles,
    autoIndex: this.autoIndex,
    defaultExt: this.ext,
    gzip: this.gzip,
    brotli: this.brotli,
    contentType: this.contentType,
    handleError: typeof options.proxy !== 'string'
  }));

  if (typeof options.proxy === 'string') {//如果 设置了代理
    var proxy = httpProxy.createProxyServer({});
    before.push(function (req, res) {
      proxy.web(req, res, {
        target: options.proxy,
        changeOrigin: true
      }, function (err, req, res, target) {
        if (options.logFn) {
          options.logFn(req, res, {
            message: err.message,
            status: res.statusCode });
        }
        res.emit('next');
      });
    });
  }

  var serverOptions = {
    before: before,
    headers: this.headers,
    onError: function (err, req, res) {
      if (options.logFn) {
        options.logFn(req, res, err);
      }

      res.end();
    }
  };

  if (options.https) { //设置 https
    serverOptions.https = options.https;
  }

  this.server = union.createServer(serverOptions);
  if (options.timeout !== undefined) {
    this.server.setTimeout(options.timeout);
  }
}

HttpServer.prototype.listen = function () {
  this.server.listen.apply(this.server, arguments);
};

HttpServer.prototype.close = function () {
  return this.server.close();
};
